scheduler/reconcile.go

```diff
@@ -961,6 +992,99 @@ func (a *allocReconciler) computeStop(group *structs.TaskGroup, nameIndex *alloc
 	return stop
 }
 
+// computeStopByReconnecting helps build the stop set by analyzing the reconnecting
+// and untainted sets. The function accepts and decrements a target number of
+// allocations to remove. If at any time that reaches zero, the function exits.
+// If there are still allocations to be removed after analyzing the full reconnecting
+// set, that count is returned.
+// The following rules are enforced.
+// * If the reconnecting alloc's desired status in not run, stop it.
+// * If the user-specified desired transition for the reconnecting alloc is not run, stop it.
+// * If the reconnecting alloc has been rescheduled, compare the node NormScores and keep the best.
+func (a *allocReconciler) computeStopByReconnecting(untainted, reconnecting, stop allocSet, remove int) int {
+	for _, reconnectingAlloc := range reconnecting {
+		// if the desired status is not run, stop the allocation.
+		if reconnectingAlloc.DesiredStatus != structs.AllocDesiredStatusRun {
+			stop[reconnectingAlloc.ID] = reconnectingAlloc
+			a.result.stop = append(a.result.stop, allocStopResult{
+				alloc:             reconnectingAlloc,
+				statusDescription: allocNotNeeded,
+			})
+			delete(reconnecting, reconnectingAlloc.ID)
+
+			remove--
+			// if we've removed all we need to, stop iterating and return.
+			if remove == 0 {
+				return remove
+			}
+		}
+
+		// if the user-specified desired transition is not run, stop the allocation.
+		if reconnectingAlloc.DesiredTransition.ShouldMigrate() ||
+			reconnectingAlloc.DesiredTransition.ShouldReschedule() ||
+			reconnectingAlloc.DesiredTransition.ShouldForceReschedule() {
```

Comment:
Yeah. I was trying to reduce iterations. I've combined the conditionals as you suggested, which is a clear improvement. If that still feels like too much after you see the update, let me know and I can break it up further. I also added a guard clause if remove == 0 and changed these `logger.Debug` to `logger.Error` since I think these are actual error conditions.

Updated in next push.